massa-models/src/endorsement.rs

```diff
@@ -119,128 +123,131 @@ pub struct Endorsement {
-/// - Validity of the slot.
-impl SerializeCompact for Endorsement {
-    fn to_bytes_compact(&self) -> Result<Vec<u8>, ModelsError> {
-        let mut res: Vec<u8> = Vec::new();
-
-        // Sender public key
-        res.extend(&self.sender_public_key.to_bytes());
-
-        // Slot
-        res.extend(self.slot.to_bytes_compact()?);
+/// Serializer for `Endorsement`
+pub struct EndorsementSerializer {
+    slot_serializer: SlotSerializer,
+    u32_serializer: U32VarIntSerializer,
+}
 
-        // endorsement index inside the block
-        res.extend(self.index.to_varint_bytes());
+impl EndorsementSerializer {
+    /// Creates a new `EndorsementSerializer`
+    pub fn new() -> Self {
+        EndorsementSerializer {
+            slot_serializer: SlotSerializer::new(),
+            u32_serializer: U32VarIntSerializer::new(),
+        }
+    }
+}
 
-        // id of endorsed block
-        res.extend(self.endorsed_block.to_bytes());
+impl Default for EndorsementSerializer {
+    fn default() -> Self {
+        Self::new()
+    }
+}
 
-        Ok(res)
+impl Serializer<Endorsement> for EndorsementSerializer {
+    fn serialize(&self, value: &Endorsement, buffer: &mut Vec<u8>) -> Result<(), SerializeError> {
+        self.slot_serializer.serialize(&value.slot, buffer)?;
+        self.u32_serializer.serialize(&value.index, buffer)?;
+        buffer.extend(value.endorsed_block.0.to_bytes());
+        Ok(())
     }
 }
 
-/// Checks performed:
-/// - Validity of the sender public key.
-/// - Validity of the slot.
-/// - Validity of the endorsement index.
-/// - Validity of the endorsed block id.
-impl DeserializeCompact for Endorsement {
-    fn from_bytes_compact(buffer: &[u8]) -> Result<(Self, usize), ModelsError> {
-        let max_block_endorsements =
-            with_serialization_context(|context| context.endorsement_count);
-        let mut cursor = 0usize;
-
-        // sender public key
-        let sender_public_key = PublicKey::from_bytes(&array_from_slice(&buffer[cursor..])?)?;
-        cursor += PUBLIC_KEY_SIZE_BYTES;
-
-        // slot
-        let (slot, delta) = Slot::from_bytes_compact(&buffer[cursor..])?;
-        if slot.period == 0 {
-            return Err(ModelsError::DeserializeError(
-                "the target period of an endorsement cannot be 0".into(),
-            ));
+/// Deserializer for `Endorsement`
+pub struct EndorsementDeserializer {
+    slot_deserializer: SlotDeserializer,
+    u32_deserializer: U32VarIntDeserializer,
+    hash_deserializer: HashDeserializer,
+}
+
+impl EndorsementDeserializer {
+    /// Creates a new `EndorsementDeserializer`
+    pub fn new(endorsement_count: u32) -> Self {
+        #[cfg(feature = "sandbox")]
+        let thread_count = *THREAD_COUNT;
+        #[cfg(not(feature = "sandbox"))]
+        let thread_count = THREAD_COUNT;
+        EndorsementDeserializer {
+            slot_deserializer: SlotDeserializer::new(
+                (Included(1), Included(u64::MAX)),
```

Comment:
I think we should start from 0 included here

Reply:
Ok I will make the change but the old code had : 
```rust
if slot.period == 0 {
            return Err(ModelsError::DeserializeError(
                "the target period of an endorsement cannot be 0".into(),
            ));
}
```